http-body-util/src/either.rs

```diff
@@ -0,0 +1,183 @@
+use std::error::Error;
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use bytes::Buf;
+use http::HeaderMap;
+use http_body::{Body, SizeHint};
+use proj::EitherProj;
+
+/// sum type with two cases: `Left` and `Right`, used if a body can be one of two distinct types.
+#[derive(Debug, Clone, Copy)]
+pub enum Either<L, R> {
+    /// A value of type `L`
+    Left(L),
+    /// A value of type `R`
+    Right(R),
+}
+
+impl<L, R> Either<L, R> {
+    pub(crate) fn project(self: Pin<&mut Self>) -> EitherProj<L, R> {
```

Comment:
I wonder if we should just use the regular pin-project library for this, since iirc it has enum support.

Reply:
I used pin-project in my first prototype (https://gist.github.com/indietyp/0f7a707ada980f69cdab957e5d2933a6), but then decided against it due to: https://github.com/hyperium/http-body/issues/59#issuecomment-1146832845, I'd be happy to switch to pin-project if that is wanted.

Reply:
Ok cool, I think then the only thing I would like to see is a comment here explaining that this is code copied from pin-project and point to the module below.